src/functions.php

```diff
@@ -12,22 +12,29 @@
  * Creates a `Promise` which resolves with the stream data buffer
  *
  * @param ReadableStreamInterface $stream
+ * @param int $maxLength
  * @return Promise\CancellablePromiseInterface Promise<string, Exception>
  */
-function buffer(ReadableStreamInterface $stream)
+function buffer(ReadableStreamInterface $stream, $maxLength = null)
 {
     // stream already ended => resolve with empty buffer
     if (!$stream->isReadable()) {
         return Promise\resolve('');
     }
 
+    $deferred = new Promise\Deferred();
     $buffer = '';
-    $bufferer = function ($data) use (&$buffer) {
+    $bufferer = function ($data) use (&$buffer, &$bufferer, $stream, $deferred, $maxLength) {
         $buffer .= $data;
+        if ($maxLength !== null && strlen($buffer) > (int)$maxLength) {
+            $deferred->reject(new \RuntimeException('Buffer exceeded maximum length'));
+        }
     };
```

Comment:
Probably micro-optimization, but what about incremental buffer size counting?

```php
$buffer = '';
$bufferLength = 0;

$bufferer = function ($data) use (&$buffer, &$bufferLength, &$bufferer, $deferred, $maxLength) {
    $buffer .= $data;

    if (null === $maxLength) {
        return;
    }

    $bufferLength += strlen($data);

    if ($bufferLength > $maxLength) {
        $deferred->reject(new \RuntimeException('Buffer exceeded maximum length'));
    }
};
```

Reply:
Talking about optimizations, `strlen()` is actually `O(1)`, so not much would be gained :-) As an alternative, you can use `isset($buffer[$maxLength])` to avoid the expensive function call :+1: